### PR TITLE
MSL: Also pack 2- and 4- element vectors when necessary.

### PR DESCRIPTION
**Pull request by cdavis5e**
This is also needed for `VK_KHR_relaxed_block_layout` support.

### CODE REVIEW

**Comment by HansKristian-ARM** [comment]
I can see 2-element vectors, but why 4? Is it legal to emit a vec4 with just 4 byte alignment?

**Comment by cdavis5e** [comment]
> I can see 2-element vectors, but why 4? Is it legal to emit a vec4 with just 4 byte alignment?

No--it would improperly straddle in that case. But it is legal to have a `u8vec4`, for example, do that.

File: spirv_msl.cpp [hunk 1]
```diff
@@ -1231,16 +1231,20 @@ bool CompilerMSL::is_member_packable(SPIRType &ib_type, uint32_t index)
 
 	auto &mbr_type = get<SPIRType>(ib_type.member_types[index]);
 
-	// Only 3-element vectors or 3-row matrices need to be packed.
-	if (mbr_type.vecsize != 3)
+	// Only vectors or 3-row matrices need to be packed.
+	if (mbr_type.vecsize == 1 || (is_matrix(mbr_type) && mbr_type.vecsize != 3))
```
Comment (HansKristian-ARM) [inline spirv_msl.cpp]: What about matrices at weird offsets?

File: reference/shaders-msl/asm/comp/relaxed-block-layout.asm.comp [hunk 2]
```diff
@@ -7,6 +7,9 @@ struct foo
 {
     uint bar;
     packed_float3 baz;
+    uchar quux;
+    packed_uchar4 blah;
+    packed_half2 wibble;
```
Comment (HansKristian-ARM) [inline reference/shaders-msl/asm/comp/relaxed-block-layout.asm.comp]: I don't see any code trying to read from these types. We need to verify it casts to unpacked types correctly.

File: spirv_msl.cpp (hunk 1 above)
Comment (HansKristian-ARM) [reply spirv_msl.cpp]: As they are sort of arrays of vectors, packing rules might ban that. I guess relaxed_block_layout only relaxes alignment for non-array/matrix types?

Comment (cdavis5e) [reply spirv_msl.cpp]: I checked the spec with and without the extension, and yeah, it only applies to vector types.

**Pull request merged**